File: Python/DAY2.py
import re

with open('External_Files/day2.txt') as f:
    ranges = f.read().split(',')
f.close()

# alternative (easier imo) regex solution
pattern1 = re.compile(r'^(\d+)\1$')
pattern2 = re.compile(r'^(\d+)\1+$')
def find(x):
    a, b = x
    a, b = int(a), int(b)
    s1 = s2 = 0
    for k in range(a,b+1):
        s = str(k)
        if pattern1.match(s): s1 += k
        if pattern2.match(s): s2 += k
    return (s1,s2)

rs = [find(x.split('-')) for x in ranges]
ans = list(map(sum,zip(*rs)))
print(ans) # PART 1: 40214376723, PART 2: 50793864718

File: Haskell/DAY2.hs
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
module DAY2 where

import Data.List.Split (splitOn)
import Text.Regex.PCRE

main :: IO()
main = do
    file <- readFile "../External_Files/day2.txt"
    print $ checkRanges (splitOn "," file)

pattern1, pattern2 :: String
pattern1 = "^(\\d+)\\1$"
pattern2 = "^(\\d+)\\1+$"

checkRanges :: [String] -> (Int,Int)
checkRanges = foldr ((<++>) . checkRange) (0,0)

(<++>) :: (Num a) => (a,a) -> (a,a) -> (a,a)
(x,y) <++> (z,w) = (x+z,y+w)

checkRange :: String -> (Int,Int)
checkRange str = (count pattern1, count pattern2)
  where
    (a,_:b) = break (== '-') str
    range = [(read a :: Int)..(read b :: Int)]
    count pt = sum (filter (\x -> show x =~ (pt :: String)) range)